_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,401 @@
+> 6. The K-Drama fan is outraged by the JavaScript alert saying their favorite K-Drama, [Crash Landing on You](https://www.imdb.com/title/tt10850932/), stinks.
+
+In this example, we have a `<script>` element and glossed over the steps for appending the element to the DOM. In reality, polluted data gets pulled into the application in various ways. Adding untrusted data in an [injection sink](https://developer.mozilla.org/en-US/docs/Web/API/Trusted_Types_API#injection_sinks) - a Web API function that allows us to add dynamic content to our applications - is a major culprit. Examples of sinks include, but aren't limited to:
+ * methods to append to the DOM such as  `innerHTML`, `outerHTML`
+ * approaches that load external resources or navigate to external sites via a URL, such as `src` or `href` for HTML elements and `url` property on styles
+ * event handlers, such as `onmouseover` and `onerror` with an invalid `src` value
+ * global functions that evaluate and/or run code, such as `eval()`, `setTimeout()`
+
+As you can see, there are many vectors for the vulnerability. Many of these sinks have legitimate use cases when building dynamic web applications. Since the sinks are necessary for web app functionality, we must use trusted data by escaping and sanitizing it.
+
+> This is still a high-level summary. The fine print is that XSS vulnerabilities can be pretty sneaky in the different ways the attack can occur.
+
+There are different XSS attacks, each with a slightly different attack vector. We'll briefly cover how three attacks work.
+
+### Stored XSS
+
+In this flavor of XSS, the attack is persisted somewhere, like in a database. We recapped _stored XSS_ in the example above, where an agitator's terrible comment with the `script` tag persists in the database and ruins someone else's day by showing the unfriendly comment in an alert.
+
+### Reflected XSS
+
+In this attack, the malicious code sneaks in through the HTTP request, usually through URL parameters. Suppose the K-Drama site takes a search term via a URL parameter, such as:
+
+```
+https://myfavekdramas.com/dramas?search=crash+landing+on+you
+```
+
+The site then takes the search terms and displays them back to the user while calling out to the back-end to run the search.
+
+But what if an agitator constructs a URL like this?
+
+```
+https://myfavekdramas.com/dramas?search=<img src=1 onerror="alert('Doh!')"/>
+```
+
+You may think that you'd never navigate to a link like that! Who would?! But let's remember, in a previous post, you _did_ click the link in your spam email to send money to your high school sweetheart. This isn't meant as a judgment; no one is immune to clicking fishy links. Also, agitators are pretty tricksy. They might use URL shorteners to obscure the risk. 
+
+### DOM-based XSS
+
+In this attack, the agitator takes advantage of Web APIs. The attack occurs entirely within the SPA, and it's pretty much identical to Reflective XSS. 
```

Comment:
```suggestion
In this attack, the agitator takes advantage of Web APIs. The attack occurs entirely within the SPA, and it's pretty much identical to _reflected XSS_. 
```
should be 'reflected' not reflective, right? I also italicized instead of capitalizing